py-polars/polars/internals/dataframe/frame.py

```diff
@@ -1431,7 +1404,7 @@ def __setitem__(
             # TODO: Use python sequence constructors
             if not _NUMPY_AVAILABLE:
                 raise ImportError("'numpy' is required for this functionality.")
-            np = numpy_mod()
+
             value = np.array(value)
```

Comment:
On
```python
if not _NUMPY_AVAILABLE:
         aise ImportError("'numpy' is required for this functionality.")
 np = numpy_mod()

value = np.array(value)
```
Would it be possible to have the lazy loader throw on access, rather than checking all across our code whether a dependency is available, and then use it if available and throwing if not? Or does that break on type annotations?

Reply:
>  Or does that break on type annotations?

Let's find out :) 
I'll have a go and see how it works & if it behaves cleanly.

Reply:
Currently it makes `mypy` miserable, but a functionally-elegant implementation is possible (and it's actually rather nice). Uses a little `importlib` black magic to dynamically proxy the missing module with one that raises a helpful exception on attribute access. Will need further thought to see how best to handle the typing lint though...

@ritchie46: if you think it's looking good as-is, I'll follow-up on this suggestion in a separate PR.